spec.md

```diff
@@ -1913,6 +1916,66 @@ The CO MUST implement the specified error recovery behavior when it encounters t
 | Operation pending for volume | 10 ABORTED | Indicates that there is a already an operation pending for the specified volume. In general the Cluster Orchestrator (CO) is responsible for ensuring that there is no more than one call "in-flight" per volume at a given time. However, in some circumstances, the CO MAY lose state (for example when the CO crashes and restarts), and MAY issue multiple calls simultaneously for the same volume. The Plugin, SHOULD handle this as gracefully as possible, and MAY return this error code to reject secondary calls. | Caller SHOULD ensure that there are no other calls pending for the specified volume, and then retry with exponential back off. |
 
 
+#### `NodeGetVolumeStats`
+
+A Node plugin MUST implement this RPC call if it has GET_VOLUME_STATS node capability.
+`NodeGetVolumeStats` RPC call returns the volume capacity stats available for the volume.
+
+If the volume is being used in `BlockVolume` mode then `used_capacity` and `available_capacity` MAY be omitted from `NodeGetVolumeStatsResponse`.
+Similarly, inode information may be omitted from `NodeGetVolumeStatsResponse` when unavailable.
+
+
+```protobuf
+message NodeGetVolumeStatsRequest {
+  // The ID of the volume. This field is REQUIRED.
+  string volume_id = 1;
+
+  // It can be any valid path where volume was previously published.
+  // It MUST be an absolute path in the root filesystem of
+  // the process serving this request.
+  // This is a REQUIRED field.
+  string volume_path = 2;
+}
+
+message NodeGetVolumeStatsResponse {
+  repeated VolumeUsage usage = 1;
+}
+
+message VolumeUsage {
+  enum Unit {
+    UNKNOWN = 0;
+    BYTES = 1;
+    INODES = 2;
+  }
+  // The available capacity in specified Unit. This field is OPTIONAL.
+  // The value of this field MUST NOT be negative.
+  int64 available = 1;
+
+  // The total capacity in specified Unit. This field is REQUIRED.
+  // The value of this field MUST NOT be negative.
+  int64 total = 2;
+
+  // The used capacity in specified Unit. This field is OPTIONAL.
+  // The value of this field MUST NOT be negative.
+  int64 used = 3;
+
+  // The units that represent this value
+  Unit unit = 4;
+}
+```
+
+##### NodeGetVolumeStats Errors
+
+If the plugin is unable to complete the `NodeGetVolumeStats` call successfully, it MUST return a non-ok gRPC code in the gRPC status.
+If the conditions defined below are encountered, the plugin MUST return the specified gRPC error code.
+The CO MUST implement the specified error recovery behavior when it encounters the gRPC error code.
+
+
+| Condition | gRPC Code | Description | Recovery Behavior |
+|-----------|-----------|-------------|-------------------|
+| Volume does not exist | 5 NOT_FOUND | Indicates that a volume corresponding to the specified `volume_id` does not exist. | Caller MUST verify that the `volume_id` is correct and that the volume is accessible and has not been deleted before retrying with exponential back off. |
```

Comment:
This is the generic error we return when specified volume does not exist. See `NodeUnpublishVolume`  for example.

Reply:
This seems like a different case than node-unpublish-volume. For example, if the volume_id is already node-unpublished at target_path, then target_path may not exist (this is an expected possibility) - the call is deemed idempotent.

For this RPC, a missing vol staged/published at target_path is an error because of the nature of the operation: the RPC isn't changing the mount state of the volume at target_path, it's trying to query the state of the volume at target_path. And if volume_id exists, but it not present at target_path .. well, that's an error that we should bubble up to the CO.

Reply:
One might argue that NOT_FOUND isn't appropriate for the case I outlined above and that we should use a different error code for such a case - I'm fine with that too.

Reply:
hmm I think the only other error code that makes sense perhaps is `FAILED_PRECONDITION` but in this case, it is not all that different from `NOT_FOUND` error because in both the cases the CO may retry.

I think I am okay with rewording the error to "Indicates that a volume corresponding to `volume_id` does not exist on specified `volume_path`"

Reply:
I have updated the comment.